packages/auth/lib/Sessions.js

```diff
@@ -1,36 +1,77 @@
-const { NoSessionError, QueryEmailMismatchError, DestroySessionError } = require('./errors')
+const kraut = require('kraut')
+const geoForIP = require('./geoForIP')
+const { newAuthError } = require('./AuthError')
+
+const DestroySessionError = newAuthError('session-destroy-failed', 'api/auth/errorDestroyingSession')
+const InitiateSessionError = newAuthError('session-init-failed', 'api/auth/session-init-failed')
+const NoSessionError = newAuthError('no-session', 'api/token/invalid')
 
 const destroySession = async (req) => {
   return new Promise((resolve, reject) => {
     req.session.destroy(error => {
       if (error) {
-        return reject(new DestroySessionError({ req, error }))
+        return reject(new DestroySessionError({ headers: req.headers, error }))
       }
       return resolve()
     })
   })
 }
 
-const sessionByToken = async ({ pgdb, token, email: emailFromQuery, ...meta }) => {
-  const Sessions = pgdb.public.sessions
-  const session = await Sessions.findOne({
-    'sess @>': { token }
+const initiateSession = async ({ req, pgdb, ipAddress, userAgent, email }) => {
+  const phrase = `${kraut.adjectives.random()} ${kraut.verbs.random()} ${kraut.nouns.random()}`
+  const { country, city } = geoForIP(ipAddress)
+  req.session.email = email
+  req.session.ip = ipAddress
+  req.session.ua = userAgent
+  if (country || city) {
+    req.session.geo = { country, city }
+  }
+  await new Promise(function (resolve, reject) {
+    req.session.save(function (error, data) {
+      if (error) {
+        return reject(new InitiateSessionError({ headers: req.headers, error }))
+      }
+      return resolve(data)
+    })
   })
+  const session = await pgdb.public.sessions.findOne({ sid: req.sessionID })
   if (!session) {
-    throw new NoSessionError({ token, emailFromQuery, ...meta })
+    throw new NoSessionError({ email })
   }
-
-  const { email } = session.sess
-  if (emailFromQuery && email !== emailFromQuery) { // emailFromQuery might be null for old links
-    throw new QueryEmailMismatchError({ token, email, emailFromQuery })
+  return {
+    session,
+    country,
+    city,
+    phrase
   }
+}
 
-  return session
+const sessionByToken = async ({ pgdb, token, email: emailFromQuery, ...meta }) => {
+  const sessions = await pgdb.query(`
+    SELECT DISTINCT
+      s.*,
+      t."expiresAt" as "tokenExpiresAt"
+    FROM
+      sessions s
+    RIGHT OUTER JOIN
```

Comment:
omit OUTER: "The words INNER and OUTER are optional in all forms. INNER is the default; LEFT, RIGHT, and FULL imply an outer join."
https://www.postgresql.org/docs/9.2/static/queries-table-expressions.html

Reply:
Why do you use a RIGHT join here? Like this you get a result if the token exists, but no session for it. And you return that "half" result claiming it's a session. This should be a [INNER] JOIN.

Reply:
i don't know anymore but I had a good reason to use it that way, I think it was about will leave it like this until we have the tests